prqlc/prqlc/src/sql/pq/postprocess.rs

```diff
@@ -152,13 +152,86 @@ struct CteSorting {
 
 impl RqFold for SortingInference<'_> {}
 
+/// Returns the CIds for the last SELECT statement in an AtomicPipeline CTE, or None if no such
+/// statement is found.
+fn find_last_select_for_cte(cte: &Cte) -> Option<&Vec<CId>> {
+    match &cte.kind {
+        CteKind::Normal(SqlRelation::AtomicPipeline(pipeline)) => {
+            pipeline.iter().rev().find_map(|transform| {
+                if let SqlTransform::Select(cids) = transform {
+                    Some(cids)
+                } else {
+                    None
+                }
+            })
+        }
+        _ => None,
+    }
+}
+
 impl PqFold for SortingInference<'_> {
     fn fold_sql_query(&mut self, query: SqlQuery) -> Result<SqlQuery> {
         let mut ctes = Vec::with_capacity(query.ctes.len());
 
         for cte in query.ctes {
             log::debug!("infer_sorts: {0:?}", cte.tid);
+
+            let last_select_before_fold = find_last_select_for_cte(&cte).cloned();
             let cte = self.fold_cte(cte)?;
+            let last_select_after_fold = find_last_select_for_cte(&cte);
+
+            // Checking if CTE folding has added some new columns to the CTE
+            match (last_select_before_fold.as_ref(), last_select_after_fold) {
+                (Some(before), Some(after)) if before != after => {
+                    let new_columns = after
+                        .iter()
+                        .filter(|cid| !before.contains(cid))
+                        .collect::<Vec<_>>();
+
+                    // if new columns are added, the relation instance needs to be updated
+                    if !new_columns.is_empty() {
+                        let mut used_new_names = HashSet::new();
+                        let mut cid_redirects_to_add = Vec::new();
+                        for old_cid in new_columns {
+                            let new_cid = self.ctx.anchor.cid.gen();
+                            let old_name = self.ctx.anchor.ensure_column_name(*old_cid).cloned();
+                            let mut new_name = old_name;
+                            if let Some(new) = &mut new_name {
+                                if used_new_names.contains(new) {
+                                    *new = self.ctx.anchor.col_name.gen();
```

Comment:
@lukapeschke if you have the patience, can we adjust the test so we hit this line too? I don't think we hit it currently

either way let's merge, feel free to ping either way